### PR TITLE
feat(react-grid): implement column chooser

### PR DESCRIPTION
**Pull request by gsobolev**

### CODE REVIEW

File: packages/dx-react-grid/docs/reference/column-chooser.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+## User reference
+
+### Properties
+
+Name | Type | Default | Description
+-----|------|---------|------------
+columns | Array&lt;[Column](grid.md#column)&gt; | | Specifies for which row object fields columns are created.
+hiddenColumns | Array&lt;string&gt; | [] | An array containing the names of the columns to be hidden.
+onHiddenColumnsChange | (nextHiddenColumns: Array&lt;string&gt;) => void | | Handles column visibility change.
+contentTemplate | (args: [ColumnChooserRootArgs](#column-chooser-root-args)) => ReactElement | | A template that renders the column chooser markup.
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/column-chooser.md]: contentTemplate with ColumnChooserRootArgs... It's quite strange

Comment (kvet) [reply packages/dx-react-grid/docs/reference/column-chooser.md]: Also seems that we should extract container and item templates for API consistency

File: packages/dx-react-grid/docs/reference/hidden-table-columns.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# HiddenTableColumns Plugin Reference
```
Comment (kvet) [inline packages/dx-react-grid/docs/reference/hidden-table-columns.md]: Outdated?

File: packages/dx-react-grid/docs/reference/column-chooser.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Grid Reference
```
Comment (dxbykov) [inline packages/dx-react-grid/docs/reference/column-chooser.md]: Why 'Grid'?

File: packages/dx-react-grid/docs/reference/column-chooser.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+## Interfaces
+
+### <a name="column-chooser-root-args"></a>ColumnChooserRootArgs
+
+Describes properties passed to the root template when it is being rendered.
+
+Field | Type | Description
+------|------|------------
+columnChooserItems | Array&lt;[ColumnChooserItem](#column-chooser-item)&gt; | An array of column chooser items.
+onColumnToggle | (columnName: string) => void | Handles column visibility change.
```
Comment (dxbykov) [inline packages/dx-react-grid/docs/reference/column-chooser.md]: Or onToggleColumn?

Comment (gsobolev) [reply packages/dx-react-grid/docs/reference/column-chooser.md]: Discussed this with our technical writer. The `onColumnToggle` is correct.

File: packages/dx-react-grid-bootstrap3/src/templates/column-chooser-panel.jsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+export const ColumnChooserPanel = ({ columnChooserItems, onColumnToggle, itemTemplate }) => (
+  <div className="list-group">
+    {columnChooserItems.map(({ column, hidden }) => React.cloneElement(
+      itemTemplate({ column, hidden, onClick: onColumnToggle }),
```
Comment (kvet) [inline packages/dx-react-grid-bootstrap3/src/templates/column-chooser-panel.jsx]: Why do you map `onColumnToggle` to `onClick` without binding arguments? It also seems that `onClick` should be renamed to `onVisibilityToggle: (toggle?: boolean) => void` and `onColumnToggle` may be named as `onColumnVisibilityToggle: (columnName, toggle?: boolean) => void`.

File: packages/dx-react-grid-bootstrap3/src/templates/column-chooser-panel.test.jsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  it('should render items with correct props', () => {
+    const itemTemplate = jest.fn(() => <div />);
+    const columnClickHandler = () => {};
+    const columnChooserItems = [
+      { column: { name: 'a', title: 'A' }, hidden: false },
+      { column: { name: 'b' }, hidden: true },
+    ];
+    mount(
+      <ColumnChooserPanel
+        columnChooserItems={columnChooserItems}
```
Comment (kvet) [inline packages/dx-react-grid-bootstrap3/src/templates/column-chooser-panel.test.jsx]: Why not simply call it `items`?

File: packages/dx-react-grid-material-ui/src/column-chooser.jsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import { ColumnChooser as ColumnChooserBase } from '@devexpress/dx-react-grid';
+import { ColumnChooserPanel } from './templates/column-chooser-panel';
+import { ColumnChooserPanelItem } from './templates/column-chooser-panel-item';
+
+const contentTemplate = props => <ColumnChooserPanel {...props} />;
+const itemTemplate = props => <ColumnChooserPanelItem {...props} />;
+
+export const ColumnChooser = props => (
+  <ColumnChooserBase
+    contentTemplate={contentTemplate}
```
Comment (kvet) [inline packages/dx-react-grid-material-ui/src/column-chooser.jsx]: Should we call it `container`? Like we call such template in DragDropContext.

File: packages/dx-react-grid/docs/guides/column-visibility.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+Assign a function that sends the `hiddenColumns` state variable to the ColumnChooser's `onHiddenColumnsChange` property to update the state once a user checks or unchecks a column item.
+
+```js
+this.hiddenColumnsChangeHandler = (hiddenColumns) => {
+  this.setState({ hiddenColumns });
+};
+```
+
+At this point, the Grid and the ColumnChooser a synchronized.
```
Comment (dirk-pieterse) [inline packages/dx-react-grid/docs/guides/column-visibility.md]: "At this point, the Grid and the ColumnChooser a synchronized."?

**Pull request merged**